### PR TITLE
Sol error types refuctoring

### PR DESCRIPTION
**Pull request by LianaHus**
For next PR:
some exceptions in ReferencesResolver should be changed
catches from test functions should be removed
some todos

### CODE REVIEW

File: libsolidity/CompilerStack.cpp [hunk 1]
```diff
@@ -97,30 +97,41 @@ void CompilerStack::setSource(string const& _sourceCode)
 
 bool CompilerStack::parse()
 {
+	// todo not sure about clear. can contain warnings
```
Comment (chriseth) [inline libsolidity/CompilerStack.cpp]: Clearing is fine, this is the first step in the whole process.

File: libsolidity/CompilerStack.cpp [hunk 2]
```diff
@@ -97,30 +97,41 @@ void CompilerStack::setSource(string const& _sourceCode)
...
 	m_errors.clear();
```
Comment (chriseth) [inline libsolidity/CompilerStack.cpp:103]: I think we should set `m_parseSuccessful = false` here.

File: libsolidity/CompilerStack.cpp [hunk 3]
```diff
@@ -97,30 +97,41 @@ void CompilerStack::setSource(string const& _sourceCode)
...
+
 	resolveImports();
 
 	m_globalContext = make_shared<GlobalContext>();
-	NameAndTypeResolver resolver(m_globalContext->declarations());
+	bool success = true;
+	NameAndTypeResolver resolver(m_globalContext->declarations(), m_errors);
 	for (Source const* source: m_sourceOrder)
-		resolver.registerDeclarations(*source->ast);
+		success = success && resolver.registerDeclarations(*source->ast);
```
Comment (chriseth) [inline libsolidity/CompilerStack.cpp]: What about `if (!resolver.registerDeclarations(*source->ast)) return false;`?
I think the current way is a bit misleading becouse of short-circuiting.

File: libsolidity/Exceptions.h [hunk 4]
```diff
@@ -31,17 +31,54 @@ namespace dev
...
-struct DeclarationError: virtual Error {};
-struct DocstringParsingError: virtual Error {};
-struct Warning: virtual Error {};
+class Error;
+using ErrorList = std::vector<std::shared_ptr<Error const>>;
 
 struct CompilerError: virtual Exception {};
 struct InternalCompilerError: virtual Exception {};
-struct FatalError: virtual Exception {};
+struct fatalError: virtual Exception {}; //todo rename to FatalError
```
Comment (chriseth) [inline libsolidity/Exceptions.h]: Yes, please rename :-)

File: libsolidity/Exceptions.h [hunk 5]
```diff
@@ -31,17 +31,54 @@ namespace dev
...
+
+	Type type() const { return m_type; }
+	std::string const& typeName() const { return m_typeName; }
+
+	/// helper functions
+	static Error const* containsErrorOfType(ErrorList const& _list, Error::Type _type)
+	{
+		for (auto e: _list)
+		{
+			if(e->type() == _type)
```
Comment (chriseth) [inline libsolidity/Exceptions.h]: Space (also below).

File: libsolidity/NameAndTypeResolver.cpp [hunk 6]
```diff
@@ -33,88 +33,114 @@ namespace solidity
...
+bool NameAndTypeResolver::updateDeclaration(Declaration const& _declaration)
 {
-	m_scopes[nullptr].registerDeclaration(_declaration, false, true);
-	solAssert(_declaration.scope() == nullptr, "Updated declaration outside global scope.");
+	try
+	{
+		m_scopes[nullptr].registerDeclaration(_declaration, false, true);
+		solAssert(_declaration.scope() == nullptr, "Updated declaration outside global scope.");
+	}
+	catch(fatalError _error)
```
Comment (chriseth) [inline libsolidity/NameAndTypeResolver.cpp]: Space

File: libsolidity/NameAndTypeResolver.cpp [hunk 7]
```diff
@@ -270,9 +293,53 @@ vector<_T const*> NameAndTypeResolver::cThreeMerge(list<list<_T const*>>& _toMer
...
-	m_scopes(_scopes), m_currentScope(nullptr)
+void NameAndTypeResolver::reportDeclarationError(
+	SourceLocation _sourceLoction,
+	string const& _description,
+	SourceLocation _secondarySourceLocation = SourceLocation(),
+	string const& _secondaryDescription = ""
+)
+{
+	auto err = make_shared<Error>(Error::Type::DeclarationError); // todo remove Error?
+	*err <<
```
Comment (chriseth) [inline libsolidity/NameAndTypeResolver.cpp:304]: I think if secondary source location is not given, we should not append it to the error, at least it makes a difference.

File: libsolidity/NameAndTypeResolver.cpp [hunk 8]
```diff
@@ -270,9 +293,53 @@ vector<_T const*> NameAndTypeResolver::cThreeMerge(list<list<_T const*>>& _toMer
...
+			errinfo_secondarySourceLocation(
+				SecondarySourceLocation().append(_secondaryDescription, _secondarySourceLocation)
+			);
+
+	m_errors.push_back(err);
+}
+
+void NameAndTypeResolver::reportFatalDeclarationError(
+	SourceLocation _sourceLoction,
+	string _description
```
Comment (chriseth) [inline libsolidity/NameAndTypeResolver.cpp]: `const&`

File: libsolidity/NameAndTypeResolver.cpp [hunk 9]
```diff
@@ -270,9 +293,53 @@ vector<_T const*> NameAndTypeResolver::cThreeMerge(list<list<_T const*>>& _toMer
...
+			errinfo_comment(_description) <<
+			errinfo_secondarySourceLocation(
+				SecondarySourceLocation().append(_secondaryDescription, _secondarySourceLocation)
+			);
+
+	m_errors.push_back(err);
+}
+
+void NameAndTypeResolver::reportFatalDeclarationError(
+	SourceLocation _sourceLoction,
```
Comment (chriseth) [inline libsolidity/NameAndTypeResolver.cpp]: Typo: `c_a_tion`

File: libsolidity/NameAndTypeResolver.cpp [hunk 10]
```diff
@@ -270,9 +293,53 @@ vector<_T const*> NameAndTypeResolver::cThreeMerge(list<list<_T const*>>& _toMer
...
+	string _description
+)
+{
+	reportDeclarationError(_sourceLoction, _description);
+	BOOST_THROW_EXCEPTION(fatalError());
+}
+
+void NameAndTypeResolver::reportTypeError(Error _e)
```
Comment (chriseth) [inline libsolidity/NameAndTypeResolver.cpp]: `const&`

File: libsolidity/NameAndTypeResolver.cpp [hunk 11]
```diff
@@ -270,9 +293,53 @@ vector<_T const*> NameAndTypeResolver::cThreeMerge(list<list<_T const*>>& _toMer
...
+{
+	m_errors.push_back(make_shared<Error>(_e));
+}
+
+void NameAndTypeResolver::reportFatalTypeError(Error _e)
```
Comment (chriseth) [inline libsolidity/NameAndTypeResolver.cpp]: `const&`

File: libsolidity/NameAndTypeResolver.cpp [hunk 12]
```diff
@@ -438,5 +503,32 @@ string DeclarationRegistrationHelper::currentCanonicalName() const
...
 
+void DeclarationRegistrationHelper::declarationError(
+	SourceLocation _sourceLoction,
+	string const& _description,
+	SourceLocation _secondarySourceLocation = SourceLocation(),
+	string const& _secondaryDescription = ""
+)
+{
+	auto err = make_shared<Error>(Error::Type::DeclarationError);
+	*err <<
```
Comment (chriseth) [inline libsolidity/NameAndTypeResolver.cpp:521]: Same comment about secondary location applies.

File: libsolidity/NameAndTypeResolver.h [hunk 13]
```diff
@@ -42,14 +42,17 @@ namespace solidity
 class NameAndTypeResolver: private boost::noncopyable
 {
 public:
-	NameAndTypeResolver(std::vector<Declaration const*> const& _globals);
+	NameAndTypeResolver(std::vector<Declaration const*> const& _globals, ErrorList& _errors);
 	/// Registers all declarations found in the source unit.
-	void registerDeclarations(SourceUnit& _sourceUnit);
+	/// @returns false in case of type error.
```
Comment (chriseth) [inline libsolidity/NameAndTypeResolver.h]: ... in case of error
(not only type error) - also below

File: libsolidity/Parser.cpp [hunk 14]
```diff
@@ -66,25 +66,35 @@ class Parser::ASTNodeFactory
 
 ASTPointer<SourceUnit> Parser::parse(shared_ptr<Scanner> const& _scanner)
 {
-	m_scanner = _scanner;
-	ASTNodeFactory nodeFactory(*this);
-	vector<ASTPointer<ASTNode>> nodes;
-	while (m_scanner->currentToken() != Token::EOS)
-	{
-		switch (auto token = m_scanner->currentToken())
+	try{
```
Comment (chriseth) [inline libsolidity/Parser.cpp]:
```
try
{
```

File: libsolidity/Parser.cpp [hunk 15]
```diff
@@ -66,25 +66,35 @@ class Parser::ASTNodeFactory
...
+			case Token::Library:
+				nodes.push_back(parseContractDefinition(token == Token::Library));
+				break;
+			default:
+				fatalParserError(std::string("Expected import directive or contract definition."));
+			}
 		}
+		return nodeFactory.createNode<SourceUnit>(nodes);
+	}
+	catch(fatalError const& _error)
```
Comment (chriseth) [inline libsolidity/Parser.cpp]: Space.

File: libsolidity/Parser.cpp [hunk 16]
```diff
@@ -1156,13 +1166,21 @@ ASTPointer<ParameterList> Parser::createEmptyParameterList()
...
 
-ParserError Parser::createParserError(string const& _description) const
+void Parser::parserError(string const& _description)
 {
-	return ParserError() <<
-		errinfo_sourceLocation(SourceLocation(position(), position(), sourceName())) <<
-		errinfo_comment(_description);
+	auto err = make_shared<Error>(Error::Type::ParserError);
+	*err <<
+			errinfo_sourceLocation(SourceLocation(position(), position(), sourceName())) <<
```
Comment (chriseth) [inline libsolidity/Parser.cpp]: Wrong indentation.

File: libsolidity/Parser.h [hunk 17]
```diff
@@ -34,7 +34,8 @@ class Scanner;
 class Parser
 {
 public:
-	Parser() {}
+	Parser(ErrorList& errors):
+		m_errors(errors){};
```
Comment (chriseth) [inline libsolidity/Parser.h]: Better in one line?

File: libsolidity/TypeChecker.cpp [hunk 18]
```diff
@@ -36,21 +36,26 @@ bool TypeChecker::checkTypeRequirements(const ContractDefinition& _contract)
...
-	bool success = true;
-	for (auto const& it: m_errors)
-		if (!dynamic_cast<Warning const*>(it.get()))
-		{
-			success = false;
-			break;
-		}
-	return success;
+
+return	Error::containsOnlyWarnings(m_errors);
```
Comment (chriseth) [inline libsolidity/TypeChecker.cpp]: Please fix this.

File: libsolidity/TypeChecker.cpp [hunk 19]
```diff
@@ -67,6 +72,7 @@ TypePointer const& TypeChecker::type(VariableDeclaration const& _variable) const
 
 bool TypeChecker::visit(ContractDefinition const& _contract)
 {
+
```
Comment (chriseth) [inline libsolidity/TypeChecker.cpp]: Unnecessary newline.

File: libsolidity/TypeChecker.h [hunk 20]
```diff
@@ -42,26 +42,26 @@ namespace solidity
 class TypeChecker: private ASTConstVisitor
 {
 public:
+	/// @_errors the reference to the list of errors and warnings to add them found during type checking.
```
Comment (chriseth) [inline libsolidity/TypeChecker.h]: I think it should be
`@param _errors the ...`

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 21]
```diff
@@ -44,18 +44,22 @@ namespace test
...
-	Parser parser;
+	ErrorList errors;
+	Parser parser(errors);
 	ASTPointer<SourceUnit> sourceUnit;
-	shared_ptr<Exception const> err;
 	// catch exceptions for a transition period
 	try
 	{
 		sourceUnit = parser.parse(std::make_shared<Scanner>(CharStream(_source)));
+		if(!sourceUnit)
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:57]: I would say it is better to use `BOOST_REQUIRE` here - there should not be any parser error when this function is called.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 22]
```diff
@@ -44,18 +44,22 @@ namespace test
...
+			return make_pair(sourceUnit, nullptr);
+
 		std::shared_ptr<GlobalContext> globalContext = make_shared<GlobalContext>();
-		NameAndTypeResolver resolver(globalContext->declarations());
+		NameAndTypeResolver resolver(globalContext->declarations(), errors);
+		solAssert(Error::containsOnlyWarnings(errors), "");
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:62]: `BOOST_REQUIRE`?

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 23]
```diff
@@ -71,51 +75,45 @@ parseAnalyseAndReturnError(string const& _source, bool _reportWarnings = false)
...
+						return make_pair(sourceUnit, std::make_shared<Error::Type const>(currentError->type()));
 				}
 			}
 	}
-	catch (ParserError const& _exception)
-	{
-		return make_pair(sourceUnit, make_shared<ParserError>(_exception));
-	}
-	catch (DeclarationError const& _exception)
+	catch(Error const& _e)
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:93]: Space.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 24]
```diff
@@ -71,51 +75,45 @@ parseAnalyseAndReturnError(string const& _source, bool _reportWarnings = false)
...
+		return make_pair(sourceUnit, nullptr);
 	}
-	return make_pair(sourceUnit, err);
+	return make_pair(sourceUnit, nullptr);
 }
 
 ASTPointer<SourceUnit> parseAndAnalyse(string const& _source)
 {
 	auto sourceAndError = parseAnalyseAndReturnError(_source);
-	BOOST_REQUIRE(!sourceAndError.second);
```
Comment (chriseth) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:110]: Can we leave this line here, too?

File: test/libsolidity/SolidityParser.cpp [hunk 25]
```diff
@@ -39,10 +39,11 @@ namespace test
 
 namespace
 {
-ASTPointer<ContractDefinition> parseText(std::string const& _source)
+ASTPointer<ContractDefinition> parseText(std::string const& _source, ErrorList& _errors)
 {
-	Parser parser;
-	ASTPointer<SourceUnit> sourceUnit = parser.parse(std::make_shared<Scanner>(CharStream(_source)));
+	ASTPointer<SourceUnit> sourceUnit = Parser(_errors).parse(std::make_shared<Scanner>(CharStream(_source)));
+	if(!sourceUnit)
```
Comment (chriseth) [inline test/libsolidity/SolidityParser.cpp]: Space

File: test/libsolidity/SolidityParser.cpp [hunk 26]
```diff
@@ -50,8 +51,30 @@ ASTPointer<ContractDefinition> parseText(std::string const& _source)
...
 
-static void checkFunctionNatspec(ASTPointer<FunctionDefinition> _function,
-								 std::string const& _expectedDoc)
+bool successParse(std::string const& _source)
+{
+	ErrorList errors;
+	try
+	{
+		auto sourceUnit = parseText(_source, errors);
+		if(!sourceUnit)
```
Comment (chriseth) [inline test/libsolidity/SolidityParser.cpp]: Space

File: test/libsolidity/SolidityParser.cpp [hunk 27]
```diff
@@ -50,8 +51,30 @@ ASTPointer<ContractDefinition> parseText(std::string const& _source)
...
+			return false;
```
Comment (chriseth) [inline test/libsolidity/SolidityParser.cpp:61]: Please `BOOST_CHECK` that `errors` contains a ParserError here.

File: test/libsolidity/SolidityParser.cpp [hunk 28]
```diff
@@ -50,8 +51,30 @@ ASTPointer<ContractDefinition> parseText(std::string const& _source)
...
+	}
+	catch (fatalError const& _exception)
+	{
+		if (Error::containsErrorOfType(errors, Error::Type::ParserError))
```
Comment (chriseth) [inline test/libsolidity/SolidityParser.cpp:65]: This should be a `BOOST_CHECK`.

File: test/libsolidity/SolidityParser.cpp [hunk 29]
```diff
@@ -50,8 +51,30 @@ ASTPointer<ContractDefinition> parseText(std::string const& _source)
...
+			return false;
+	}
+	if (Error::containsErrorOfType(errors, Error::Type::ParserError))
+		return false;
+
```
Comment (chriseth) [inline test/libsolidity/SolidityParser.cpp:70]: Please `BOOST_CHECK` that errors contains only warnings.

File: test/libsolidity/SolidityParser.cpp [hunk 30]
```diff
@@ -145,7 +168,8 @@ BOOST_AUTO_TEST_CASE(two_exact_functions)
 	// with support of overloaded functions, during parsing,
 	// we can't determine whether they match exactly, however
 	// it will throw DeclarationError in following stage.
-	BOOST_CHECK_NO_THROW(parseText(text));
+	// TODO add test to the SolidityNameAndTypeDeclaration
```
Comment (chriseth) [inline test/libsolidity/SolidityParser.cpp]: What does this todo mean?

File: test/libsolidity/SolidityParser.cpp [hunk 31]
```diff
@@ -813,10 +845,10 @@ BOOST_AUTO_TEST_CASE(external_variable)
 		contract c {
 			uint external x;
 		})";
-	BOOST_CHECK_THROW(parseText(text), ParserError);
+	BOOST_CHECK(!successParse(text));
 }
 
-BOOST_AUTO_TEST_CASE(arrays_in_storage)
+BOOST_AUTO_TEST_CASE(arrays_in_storagze)
```
Comment (chriseth) [inline test/libsolidity/SolidityParser.cpp]: Typo

File: test/libsolidity/solidityExecutionFramework.h [hunk 32]
```diff
@@ -67,12 +67,26 @@ class ExecutionFramework
...
 		m_compiler.reset(false, m_addStandardSources);
 		m_compiler.addSource("", _sourceCode);
-		BOOST_REQUIRE_THROW(m_compiler.compile(m_optimize, m_optimizeRuns), Exceptiontype);
+		bool foundError = false;
+		try
+		{
+			m_compiler.compile(m_optimize, m_optimizeRuns);
+			BOOST_REQUIRE(Error::containsErrorOfType(m_compiler.errors(), _type));
+		}
+		catch(Error const& _e)
```
Comment (chriseth) [inline test/libsolidity/solidityExecutionFramework.h]: Space

File: test/libsolidity/solidityExecutionFramework.h [hunk 33]
```diff
@@ -67,12 +67,26 @@ class ExecutionFramework
...
+		{
+			BOOST_REQUIRE(_e.type() == _type);
+			foundError = true;
+		}
+		catch(Exception const& _exception)
```
Comment (chriseth) [inline test/libsolidity/solidityExecutionFramework.h]: Space

File: libsolidity/NameAndTypeResolver.cpp [hunk 34]
```diff
@@ -270,9 +293,53 @@ vector<_T const*> NameAndTypeResolver::cThreeMerge(list<list<_T const*>>& _toMer
...
+	string const& _secondaryDescription = ""
+)
+{
+	auto err = make_shared<Error>(Error::Type::DeclarationError); // todo remove Error?
+	*err <<
+			errinfo_sourceLocation(_sourceLoction) <<
+			errinfo_comment(_description) <<
+			errinfo_secondarySourceLocation(
+				SecondarySourceLocation().append(_secondaryDescription, _secondarySourceLocation)
+			);
```
Comment (LianaHus) [inline libsolidity/NameAndTypeResolver.cpp]: indent

File: libsolidity/Parser.cpp [hunk 35]
```diff
@@ -1156,13 +1166,21 @@ ASTPointer<ParameterList> Parser::createEmptyParameterList()
...
+			errinfo_comment(_description);
```
Comment (LianaHus) [inline libsolidity/Parser.cpp]: indent

**Pull request merged**